### PR TITLE
Add Hyperopt to User & Developer Guide

### PR DESCRIPTION
**Pull request by kaushikb11**
For #677 

#### Add Hyperopt to [User Guide](https://uber.github.io/ludwig/user_guide/) with documentation that explains on how to use it and to [Developers guide](https://uber.github.io/ludwig/developer_guide/), that explains the design and the Strategy and executor abstract classes and explains how to implement a new strategy or a new executor.

### CODE REVIEW

**Comment by w4nderlust** [comment]
I added some comments on the user guide, cleaned up the developer guide part that needed a bit more work, and also added some todos in the developer guide

**Comment by kaushikb11** [comment]
Got it. Will do the required changes. Thanks

**Comment by w4nderlust** [comment]
Oh sorry, the comments are not in the my, but if you scroll up here you'll find a bunch.
Regarding the examples, usage examples are always welcome, you can show how the function is called by providing a plausible input and showing what would be a likely output (if the function returns something).

**Comment by kaushikb11** [comment]
@w4nderlust : Hi, Should I update the root key `measure` to `metric` in the functionalities? There are instances where the `metric` keyword is used as the input parameter for the classes.

File: mkdocs/docs/user_guide.md [hunk 1]
```diff
@@ -528,6 +529,286 @@ optional arguments:
...
+#### Float parameters
+
+For a `float` value, the parameters to specify are:
+- `low`: minimum value
+- `high`: maximum value
+- `scale`: `log` or `linear`
+- `steps`: OPTIONAL number of steps.
+For instance low:0.0, high:1.0, steps 3 would yield [0.0, 0.5, 1.0] for the search
+
+In the future the distribution to use (`uniform`, `gaussian`, `gamma`, etc.) with the distribution parameters could be specified, but to begin with just those parameters are sufficient.
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply mkdocs/docs/user_guide.md]: I don't think I have the permission to add a task on the Projects page. :sweat_smile:

[the remark this answers is not in the document]
Comment (w4nderlust) [reply mkdocs/docs/user_guide.md]: I will do it no worries :)

**Comment by w4nderlust** [comment]
I'm doing a pass on the docs, but I noticed one thing that could be an error: in the parallel executor, when `gpus` is `None` we should check if gpus are available (there's a TF function for that) and use that instead to determine the available gpus first, if there are none, then our logic continues correctly. So we should add this additional logic before line 501.

**Comment by w4nderlust** [comment]
Also noticed that we were not updating the strategy at end of the batch! I updated the code f all executors to do it, bu please doublecheck  if it is fine for you too.

**Comment by kaushikb11** [comment]
> @kaushikb11 Also noticed that we were not updating the strategy at end of the batch! I updated the code f all executors to do it, bu please doublecheck if it is fine for you too.

@w4nderlust I went through the code and ran it, the functionality makes sense. It's great how the code gets refined after writing tests and docs. But I didn't understand the idea of using class variables where we already have default values for the parameters.

**Comment by w4nderlust** [comment]
I'm using this pattern also in the TF2 porting. The point is that if someone wants to contribute a new executor or strategy, they don't have to do and put the defaults parameters in another different function, but that function that populates the defaults can extract the parameters from the newly added class. It's much cleaner, and if, for instance, I want to change the default parameter for num workers, I have to change it only in the ParallelExecutor class and not also in the function that populates the defaults.

**Comment by w4nderlust** [comment]
Another thing: I noticed only now that psutils is an external library that needs to be added to the requirements. There's no way to obtain the same result without using it, just with functions from the standard library?

**Comment by kaushikb11** [comment]
> Another thing: I noticed only now that psutils is an external library that needs to be added to the requirements. There's no way to obtain the same result without using it, just with functions from the standard library?

I found something interesting while searching for it. In `scikit-learn`,`n_jobs` input param for the Parameter Search Classes has the following behavior:
> n_jobs is an integer, specifying the maximum number of concurrently running workers. If 1 is given, no joblib parallelism is used at all, which is useful for debugging. If set to -1, all CPUs are used. For n_jobs below -1, (n_cpus + 1 + n_jobs) are used. For example with n_jobs=-2, all CPUs but one are used.

It uses the `joblib` library to run the tasks parallelly. [Link](https://github.com/joblib/joblib/blob/e959a01e9f0c160e9e89f6892928d755d02f2065/joblib/externals/loky/backend/context.py#L104) on how they figure out the number of CPUs the current process can use. The only downside of using `os.sched_getaffinity(0)` is that it appears to be UNIX only. But they have added Exception Handling and get `aggregated_cpu_count` by taking the minimum of the implemented methods.

**Comment by w4nderlust** [comment]
Does this account for the logical cpus vs physical cpus?

**Comment by kaushikb11** [comment]
> Does this account for the logical cpus vs physical cpus?

They are considering logical cores as well. I am running an experiment on multiprocessing for data-intensive tasks. I am training ten instances of models using 4 cores (`mp.cpu_count`) and 2 cores (`psutil.cpu_count(logical=False)`)

**Comment by w4nderlust** [comment]
It kinda makes sense right? The logical cores spend more time swapping stuff in and out so they are less utilized, while the physical cores don't have to do it so are more fully utilized. Because of this anyway, I have the feeling we can safely remove the check for physical vs logical, and consider only logical. My original fear is that if we consider logical, the performance would have been degraded, but your test shows that, although it doesn't increase  much, at least it does not degrade. So I would say let's remove psutils to get the number of cores and let's use something in the standard library, even if it returns the number of logical cores instead of physical (and let's make sure it works on all platforms). Does that sound good?

**Comment by kaushikb11** [comment]
Sounds good to me!

**Pull request merged**